### ISSUE TITLE
qTouch

### ISSUE THREAD

**Issue by MX682X** [SpenceKonde/AVR_Research#2]
my biggest pet peeve so far is the missing ptc support....
So I've looked through the assembly, together with ghidra and wrote a basic library to support selfcap and mutual cap buttons.
It is based on the qTouch Modular library, but I tried to do it differently, as much as reasonable, to avoid getting smited by MC (better safe then sorry). Furthermore I'm not gonna do the other stuff like sliders and 2D surfaces.
I've also realized that qTouch does a lot of pointer arithmethic due to the structure of the struct types. By avoiding that I was able to shrink the library size aswell (about 50% I guess).
Also posting it here for now because:
A. It has research about the PTC registers
B. Trying to keep it off the radar for now.
[sketch_apr04a-002.zip](https://github.com/SpenceKonde/AVR_Research/files/11186268/sketch_apr04a-002.zip)

P.S.: Fun fact: The DA has actually two ADCs - the second is inside the PTC module - that's why the Tiny PTC "Takes over ADC0" but this restriction does not appy to DA. Not so fun Fact: As the register layout is different, the DA will need it's own library... (the 328PB is V1, the Tiny 1-Series is V2, the DA is V3)

**Comment by SpenceKonde** [comment]
Downloaded. If I were you, I;d delete that for now. I'd try to trick someone someone else into releasing it - unless, of course, Microchip okays it. I could obliquely askmy microchip guy how the company would feel about that

Because they are fairly permissive. Like MicroUPDI firmware? That mEDBG, except that opened it in a hex editor and looked for an STS pointed at ADCSCRA (there was one, stored the value from register 16 there? Could be anyhing but.. what's this right before it? A load immediate to r16! So I just changed that value to it set the ADC to use Vdd as reference and we were rockin and rolling, and I DIDN'T have to make a mod fom hell between two pins of the QFN (There are a lot of parts I don't even bat an eye about, but 0.4mm pitch DFN, and then you need to wire on a jumper? No thanks. And microchip doesnt care about that

But they might care about QTouch. I've tried to reason out why it was secret

Either:
1) The code quality is so poor it would reflect poorly on the company (am told why they do not opensource the *edbgs. Then again the released what would become Serial UPDI with defalt settings that were under 1 k/s. 
2) The code would reveal too much to their competitors about the state of their cap touch sensing (I guess if they couldn't characterize any in the lab, which is hard to belive)
3) QTouch contains pieces copyrighted and licensed from others. 
4) QTouch would contain too much information about the internal implementation, not that this raises the bar for skill level needed to reverse engineer it all that much, and I question whether there's a swarm of competitors trying to reverse engineer qtouch

One thing worth noting is that @ladyada did this for the SAM parts, ro so I've been told (never owned any SAM parts. I have a board with no chip though:-(

**Comment by SpenceKonde** [comment]
but a cleanly reimplemented oner would be awesome imo better than getting a way to use the stock qtouch librries. I agree totally nobody is gong to be using for 2d surface. I do think people would want to do slider. We all know that Microchip's code quality standards leave something to be desired. 

My biggest concern is how many of the latest parts haven't had it I wonder if it's on the Microchip Shit List next to external crystals..... (which they only seem to support when s major customer pistol whip's his sales rep and demands it) Has anything since the DA had it? (ie,does educated fuzzing of the relevant registers on parts that don't advertise it to shit?)

And yeah I'd deduced that the PTC involved a second ADC, I knew there was more to the 1+ series getting dual ADCs///

**Comment by MX682X** [comment]
I forgot to add, the license.h that comes with the atmel start generated code, only forbids to use the software with non-MC parts. Nothing about forbidding decompilation, so should be technically alright, but I'll try to stick with a "trial" version. And the most interesting thing in my opinion, the harware capacitor compensation, is already documented here:
https://www.microchipdeveloper.com/touch:guide-to-interpret-cc-calibration-value

**Comment by MX682X** [comment]
My biggest concern is how many of the latest parts haven't had it 

I think this is becaus ethey do a LOT of stuff for the PTC in hardware. And with the OPAMP and PGAs, they just didn't had the space anymore. Furthermore, I think they don't use DA, DB etc. as generation disambiguation but more of a feature set. each family has something the others don't. 

Also Also, there might be a good chance that not many people were intrested in the touch functionality, especially since it required Atmel Start. It was always a hassle. And maybe there were some legal background rules that didn't allow Atmel ant then MC to publish anything after they bought the stuff.

**Comment by SpenceKonde** [comment]
And if you decompile that refactor so it's actually efficient, it will be essentially unrecognizable as their code ;-)

**Comment by ladyada** [comment]
yes i re-engineered the QTouch for SAMD21 implementation - they left some debug symbols in and it was a pretty straightforward reverse engineer. nobody from MCP has ever cared or contacted us aobut it.
https://www.youtube.com/watch?v=T1EgC_NO6qs

**Comment by MX682X** [comment]
Alright then... I guess there shouldn't be much troubple adding this to megaTinyCore.
Features: 
- Mutual- and Self-Cap support (switchable at runtime)
-  auto calibration
-  pseudo-simultanious use of ADC0. 
- Works mostly "out-of-the-box" but supports similar adjusting options you can find in the qTouch-configurator.
- Samller Memory footprint compared to qTouch as it was written for the AVR CPU, core and gcc pointer arithmetic shortcomings  in mind (e.g. use of _fastPtr, staticly linked list, 8-Bit variables preferred)

Features to be added in the future: 
- Self-cap with shield
- combining multiple buttons
-  use RTC and Event-System to allow wake-up from sleep.

I would prefer to add it to the megaTinyCore libraries, because this library only works on the 1-Series chips and because I have some pin lookup-tables that would fit perfectly into the pins_arduino.h file. Will be labeled as experimental in the Refs for now. I would make a PR if you don't mind it, Spence. 
Oh, and as I only have the 1614s, I was only able to test it there, maybe you could test it on the bigger ones aswell? Should be easily portable though.
[sketch_apr04a-008.zip](https://github.com/SpenceKonde/AVR_Research/files/11300919/sketch_apr04a-008.zip)

**Comment by SpenceKonde** [comment]
wait did you pull this off? 

O_O

**Comment by MX682X** [comment]
Well, I don't have anything explicit from MCP saying I was allowed to do what I did, but there was nothing that would have forbidden me from doing it (I didn't have to agree to anything to obtain the files), so it should be fine.

And the testing is only needed on the x16 and x17 parts. No other Tiny Family has PTC

**Comment by ladyada** [comment]
neat :) @caternuson

**Comment by SpenceKonde** [comment]
> Well, I don't have anything explicit from MCP saying I was allowed to do what I did, but there was nothing that would have forbidden me from doing it (I didn't have to agree to anything to obtain the files), so it should be fine. besides, I'm targeting a "fair use" by providing only some functionaliy of the original qtouch library
> 
> And the testing is only needed on the x16 and x17 parts. No other Tiny Family has PTC

email me your address and I'll make some parts for testing appear (note: the freebies you get for testing are the ones that are deemed non-salable for some reason, either because they're on an old rev of the board, or because they have had pins soldered onto them.)
